content/blog/how-i-learned-to-stop-worrying-and-love-performance-regression-tests.asciidoc

```diff
@@ -0,0 +1,424 @@
+
+dev.morling.demos.quarkus.TodoResourcePerformanceTest-createTodo.jfr
+dev.morling.demos.quarkus.TodoResourcePerformanceTest-retrieveTodo.jfr
+----
+
+Let's open the *.jfr file for the failing test in JDK Mission Control (JMC) in order to analyse all the recorded events
+(note that the recording will always contain some JfrUnit-internal events which are needed for synchronizing the recording stream and the events exposed to the test).
+
+When taking a look at the TLAB events of the application's executor thread,
+the culprit is identified quickly;
+a lot of the sampled TLAB allocations contain this stack trace:
+
+image::continuous_perf_testing_tlab_in_jmc.png[TLAB allocations in JDK Mission Control]
+
+Interesting, REST Assured loading a Jackson object mapper, what's going on there?
+Here's the full stacktrace:
+
+image::continuous_perf_testing_tlab_stacktrace.png[Complete stacktrace of the TLAB allocation]
+
+So it seems a REST call to another service is made from within the `TodoResource#get(long)` method!
+At this point we know where to look into the source code of the application:
+
+[source,java]
+----
+@GET
+@Transactional
+@Produces(MediaType.APPLICATION_JSON)
+@Path("/{id}")
+public Response get(@PathParam("id") long id) throws Exception {
+  Todo res = Todo.findById(id);
+  
+  User user = RestAssured.given().port(8082)
+      .when()
+          .get("/users/" + res.userId)
+          .as(User.class);
+
+  res.userName = user.name;
+
+  return Response.ok()
+      .entity(res)
+      .build();
+}
+----
+
+It seems a developer on the team has been taking the microservices mantra a bit too far and invokes another service in order to obtain some additional data associated to the user who created the retrieved todo.
+
+While that's problematic on its own right due to the inherent coupling between the two services
+(how should the todo service react if the user service isn't available?),
+they made matters worse by using the https://rest-assured.io/[REST Assured API] as a REST client in a less than ideal way.
+The APIs simplicity and elegance makes it a great choice for testing (and indeed that's its primary use case),
+but this particular usage seems to be not such a good choice for production code.
+
+At this point you should ask yourself whether the increased allocation per request actually is a problem for your application or not.
+For that it helps to run some tests on actual request latency and through-put on a production like environment.
+If there's no impact based on the workload you have to process,
+you might very well decide that additional allocations are well spent for your application's purposes.
+
+Increasing the allocation per request by a factor of 10 quite likely does not fall into this category, though.
+At the very least, we should look into making the call of the User REST API in a more efficient way,
+either by setting up REST Assured in a more suitable way, or by looking for an alternative REST client.
```

Comment:
... for an *alternate* (?) REST client.

Reply:
I think alternative actually is correct here. alternate as adverb means "wechselnd".